Add initial framework for TKW

martin-luecke opened the pull request:
TODO: expand PR description, this is currently for enabling commenting 

Detailed description of this project in [shark_turbine/kernel/functional/README.md](https://github.com/iree-org/iree-turbine/compare/wave_kernel_language?expand=1#diff-8bc4de72a74936137a9fc8cb3d47f17bb27c997ce651b876143d731a17707041)

harsh-nod commented on shark_turbine/kernel/_support/nodes.py:
```diff
@@ -0,0 +1,151 @@
+
+    def __str__(self) -> str:
+        name = get_node_name(self.__class__.__name__)
+        vars_list = [f"{key}={value}" for key, value in vars(self).items()][2:]
```
where does the number 2 come from?

harsh-nod commented on shark_turbine/kernel/_support/nodes.py:
```diff
@@ -0,0 +1,151 @@
+    op: Any
+
+    @classmethod
+    def from_fx_node(cls, node: fx.Node):
```
could you annotate the return type as well?

harsh-nod commented on shark_turbine/kernel/_support/nodes.py:
```diff
@@ -0,0 +1,151 @@
+        # If a subclass does not define custom printing we revert to the default
+        return str(self)
+
+    def add_to_graph(self, graph: fx.Graph):
```
This function is a bit confusing. When adding a node to a graph, I would have expected a graph.create_node. Instead there is a graph.create_proxy here. What is the difference? Maybe consider calling this add_proxy?

harsh-nod commented on shark_turbine/kernel/_support/nodes.py:
```diff
@@ -0,0 +1,151 @@
+        )
+
+    @classmethod
+    def handle(cls, graph, *args, **kwargs):
```
nit: return type annotation

harsh-nod commented on shark_turbine/kernel/_support/nodes.py:
```diff
@@ -0,0 +1,151 @@
+    kwargs: dict[Any, Any]
+
+    @classmethod
+    def from_fx_node(cls, node: fx.Node):
```
nit: return types here and below

stellaraccident commented:
I saw this branch in my latest fetch and was intrigued. Looks interesting

harsh-nod commented on shark_turbine/kernel/_support/tracing.py:
```diff
@@ -177,6 +178,29 @@ def __init__(self, region_graph: RegionGraph, *, grid_type: Type[Grid]):
+            # of similar nodes. We are only interested in the name prefix.
+
+            # TODO: Instead simply shape the suffix off, as this depends on the
+            #       iteration order for nodes sharing a prefic,
```
typo: prefix

harsh-nod commented on shark_turbine/kernel/functional/README.md:
```diff
@@ -0,0 +1,99 @@
+# TKF: A Wave Kernel Language for GPUs
+
+TKF is a high-level programming language designed to simplify the development of GPU micro-kernels by abstracting over intricate details of GPU hardware. It allows developers to write efficient micro-kernels, focusing on core computations while inferring the required data transfers and indexing automatically. By leveraging a wave-based programming model, TKF enables expressing programs leveraging ...
```
Could you modify this to include the mma intrinsics as well , like - TKF enables expressing programs leveraging coalesced memory accesses and matrix multiplication intrinsics effortlessly.

harsh-nod commented on shark_turbine/kernel/functional/README.md:
```diff
@@ -0,0 +1,99 @@
+        return c_reg
+
+    # repeat represents the results of the loop
+    tkf.write(repeat, c, elements_per_thread=STORE_ELEMS_PER_THREAD)
```
A minor syntax issue: here it seems like we have just defined a repeat function and are passing that function to tkf.write. Instead we might want to call repeat and get a result and then pass that to tkf.write. Something like
```
result = repeat(c_reg)
tkf.write(result, c, ...)
```
What do you think?

harsh-nod commented on shark_turbine/kernel/functional/README.md:
```diff
@@ -0,0 +1,99 @@
+def gemm(
+    a: tkl.Memory[M, K, ADDRESS_SPACE, tkl.f16],
+    b: tkl.Memory[N, K, ADDRESS_SPACE, tkl.f16],
+    c: tkl.Memory[M, N, ADDRESS_SPACE, tkl.f32],
```
One thing that also comes up is what are these M, N and K values? Are they the original tensor shapes or just the shapes the wave sees? Can you add a comment here explaining that?

harsh-nod commented on shark_turbine/kernel/functional/README.md:
````diff
@@ -0,0 +1,99 @@
+
+## Gemm example
+
+```Python
````
Is there a way to have this reference the file directly?Then you won't have to copy it here and maintain 2 versions.

harsh-nod reviewed (changes requested):
Thanks for pushing this. Just some minor comments but looks good overall.

stellaraccident commented:
One naming nit: aside from some functional inspiration, the big idea here is a wave level programming model. If we're allocating the third letter of the prefix to what something is, probably tkw vs tkf. I doubt this will be the only thing that takes more of a functional view of things. And in general, that is a how, not a what.

martin-luecke reviewed (commented):
Thank you for the comments, they should now all be addressed

martin-luecke commented:
I updated the IREE version [here](https://github.com/iree-org/iree-turbine/blob/fa8819b9a2ac446933421afeda26d58781db2dc1/iree-requirements.txt) but the CI still seems to pick up a version which does not expose the amdgpu dialect. Do you know where to adjust this? @harsh-nod @stellaraccident

stellaraccident commented:
Look to the failing CI run?

martin-luecke commented:
Sorry, I rushed a bit to get to some other work.

I only now understood that the CI pulls nod-ai/SHARK and uses those requirements.
The CI here contains an sed command to adjust the SHARK requirements from iree-turbine/main to the currently tested commit. 
The sed command is here: 

https://github.com/iree-org/iree-turbine/blob/45f1de86776c64737d8b4b8509048a0948e98177/.github/workflows/test_shark.yml#L49-L49

and seems to be supposed to match [here](https://github.com/nod-ai/SHARK/blob/78160b82e860f65748d523d9907c20ca5be44fa5/requirements.txt#L9)

However, it does not. I am not sure whether that is intended to test everything with the requirements of iree-turbine/main or an oversight. 
I modified it to do the replacement, but that currently messes up torch dependencies. Also it still seems to not pull the iree version indicated in `iree-requirements.txt`. I assumed this is the place to modify the requirement, similar to how it is done in [this patch](https://github.com/iree-org/iree-turbine/pull/8)
But I also see that in earlier CI runs unrelated to this PR a different iree version from the `iree-requirements.txt` is used for testing. e.g. [here](https://github.com/iree-org/iree-turbine/actions/runs/9088680117/job/24978647677#step:5:182) which pulls version 20240410.859 instead of version 20240514.893 indicated in the iree-requirements.

I will have to revisit this later.

harsh-nod commented on shark_turbine/kernel/_support/nodes.py:
```diff
@@ -0,0 +1,159 @@
+class ReadNode(CustomNode):
+    memory: fx.Proxy
+    elements_per_thread: Optional[Any] = None
+    type: Optional[Type[Register]] = None
```
Does Optional here mean we can omit the type?

harsh-nod commented on shark_turbine/kernel/wave/ops.py:
```diff
@@ -0,0 +1,36 @@
+    "reduction",
+]
+
+
```
Maybe not in this PR, but could these be automatically generated?

harsh-nod replied:
great, glad to know this is possible

harsh-nod commented on shark_turbine/kernel/wave/wave.py:
```diff
@@ -0,0 +1,91 @@
+        region_graph = KernelRegionGraph()
+        with CompiledContext(region_graph, grid_type=self.grid_type) as context:
+            custom_ops: dict[str, Type[CustomNode]] = {
+                "construct_register_from_metadata": ConstructRegisterFromMetadataNode,
```
Do we need to specify all new nodes here or can they be inferred from one single source of truth?

martin-luecke replied:
I reworked how new operations are registered. The single source of truth is `shark_turbine/kernel/ops/wave_ops`, e.g. [here](
https://github.com/iree-org/iree-turbine/blob/ec5b437ed1922cb6d0aea1e121d6ddd86cc3f307/shark_turbine/kernel/ops/wave_ops.py#L182)

This generates the functions used to construct and trace a program. The current limitation is proper IDE support as these functions are not statically known. I intend to work on that next and explore using `.pyi` stubs for that.

harsh-nod commented on shark_turbine/kernel/wave/wave.py:
```diff
@@ -0,0 +1,91 @@
+                root_name, _ = subtracer.trace(self._f)
+                trace = CapturedTrace(region_graph, root_name)
+
+            if dump:
```
What do you think of replace dump with a "LLVM_DEBUG" style flag that can be activated based on an incoming argument (like debug-only=)

martin-luecke replied:
I changed this to use an environment variable `DEBUG`

martin-luecke commented on shark_turbine/kernel/_support/tracing.py:
```diff
@@ -177,6 +178,29 @@ def __init__(self, region_graph: RegionGraph, *, grid_type: Type[Grid]):
+            # TODO: Instead simply shape the suffix off, as this depends on the
+            #       iteration order for nodes sharing a prefix,
+            #       e.g. read and read_shared
+            if node.name.startswith(name):
```
I agree 👍 
I added a field `tkw_op` where the custom op that was used to create this node is recorded

harsh-nod replied:
awesome thank you!

martin-luecke reviewed (commented):
Thank you for the comments @harsh-nod 
I will explore tomorrow autogenerating more of the operation boilerplate and in particular see whether the registration of the custom ops to the context can be done differently.

martin-luecke commented on shark_turbine/kernel/_support/nodes.py:
```diff
@@ -0,0 +1,159 @@
+    graph: fx.Graph
+    op: Any
+
+    @classmethod
```
I added unit tests for the tracing of all our ops

martin-luecke commented:
Oops, seems my VSCode plugin glitched and I accidentally closed. Now reopened. 
I pushed the lit tests to their own top-level directory as suggested to avoid test tools fighting each other.

martin-luecke commented on shark_turbine/kernel/ops/wave_ops.py:
```diff
@@ -0,0 +1,238 @@
+from functools import wraps
+import sys
+from typing import (
+    TYPE_CHECKING,
```
This is a constant that evaluates to `True` only when the type checker runs (e.g. mypy / Pylance). Just below I import something guarded by this constant to enable proper type checking. 
The unguarded import would result in a circular dependency during runtime, but as I need the import only for type checking this works nicely.

martin-luecke commented on tests/lit.cfg.py:
```diff
@@ -0,0 +1,61 @@
+if filecheck_exe is None:
+    filecheck_exe = shutil.which("FileCheck")
+    if filecheck_exe:
+        print(f"Using LLVM FileCheck: {filecheck_exe}")
```
Good suggestion, done! 
For wave.py I removed printing entirely as I don't think we always want to have the graph in the debug output. Initially the tests depended on that, but I found a better approach using a different decorator `wave_trace_only`

martin-luecke commented on shark_turbine/kernel/ops/wave_ops.py:
```diff
@@ -0,0 +1,238 @@
+from .base import OpDispatcher
+
+
+def get_node_name(string: str, skip_first: bool = True):
```
The fx node name depends on previous nodes in the graph, e.g. it might be `read_6`. This function extracted a readable name from our node classes. I removed it as we already give ops a name in the new `define_op` [here](https://github.com/iree-org/iree-turbine/blob/da97c1c9de86daa557902f0b0560a2f49301aa85/shark_turbine/kernel/ops/wave_ops.py#L214) and can simply use that instead.

martin-luecke commented on tests/kernel/wave/tracing.py:
```diff
@@ -0,0 +1,189 @@
+        pass
+
+    trace = test()
+    print(trace.get_root_graph())
```
I added a helper that prints all subgraphs of a trace. So all tests now simply call 
```
trace = test()
print_trace(trace)
```
and then have a bunch of CHECK statements.

martin-luecke commented:
Thank you @harsh-nod 

Inspired by your comment I added the `add_to_graph` to our custom ops and took the time to simplify the interface for adding them to the graph after tracing.
In particular I removed the `fx_op` field from CustomOp which was required for initialization before. It is more a technical detail and remnant from the previous implementation. Instead, there is now a `_tracing_function` field on the class, which is set automatically by the `define_op` decorater.

With that the API for creating nodes after the tracing looks like [this](https://github.com/iree-org/iree-turbine/blob/da97c1c9de86daa557902f0b0560a2f49301aa85/lit_tests/kernel/wave/tracing.py#L70):
```
[...]
read = Read(a).add_to_graph(graph)
write = Write(read, a, 4).add_to_graph(graph)
```

Groverkss commented on shark_turbine/kernel/lang/functional_types.py:
```diff
@@ -0,0 +1,137 @@
+        return cast(Type[MemoryTypeT], MemoryType)
+
+
+class Memory(metaclass=_MemoryStorage):
```
Is this type different from KernelBuffer? If not, can we reuse them?

harsh-nod replied:
I believe the biggest difference is the addition of the AddressSpace in the type to be able to model different levels in the memory hierarchy (think AIE for example which has multiple levels) and also didn't want to break things in the other pipeline but will let Martin comment as well.

martin-luecke replied:
Indeed, the plan is to infer where the memory is to be placed and also whether it is used for input / output / both.
Examples of the supported usage is [here](https://github.com/iree-org/iree-turbine/blob/eaa04e20389df28304eeb3efe12acc8c68738ca8/tests/kernel/wave/types_test.py#L15)

Groverkss replied:
It's fine to land this, but effectively, KernelBuffer and this MemoryStorage are the same thing. KernelBufferUsage.Local would just get another attribute specifying what kind it is.

martin-luecke replied:
You are right, I reworked the types to build on top of `KernelBufferMeta` instead.
For that I changed it as follows:
- made it public
- added address_space to it + moved enum to this file
- added the option to manually specify a type name besides inferring from KernelBufferUsage

We should only deviate from the established framework if we have good reasons, so thanks for highlighting this.

Let me know if this is alright for you.

harsh-nod commented:
> Thank you @harsh-nod
> 
> Inspired by your comment I added the `add_to_graph` to our custom ops and took the time to simplify the interface for adding them to the graph after tracing. In particular I removed the `fx_op` field from CustomOp which was required for initialization before. It is more a technical detail and remnant from the previous implementation. Instead, there is now a `_tracing_function` field on the class, which is set automatically by the `define_op` decorater.
> 
> With that the API for creating nodes after the tracing looks like [this](https://github.com/iree-org/iree-turbine/blob/da97c1c9de86daa557902f0b0560a2f49301aa85/lit_tests/kernel/wave/tracing.py#L70):
> 
> ```
> [...]
> read = Read(a).add_to_graph(graph)
> write = Write(read, a, 4).add_to_graph(graph)
> ```

Thanks Martin, that's awesome. Thank you for the push and this is a great first PR to build on top of.

harsh-nod reviewed (approved):
lgtm! thanks!

harsh-nod commented on shark_turbine/kernel/lang/functional_types.py:
```diff
@@ -0,0 +1,136 @@
+from typing import Optional, Type, TypeVar, cast, ClassVar
```
Good catch! Yeah this should be wave_types since we have renamed from functional -> wave.

martin-luecke replied:
I updated this to `wave_types.py` and added unit tests.
